### ISSUE TITLE
Wrong type inference recursively iterating union

### ISSUE THREAD

**Issue by memeplex** [microsoft/pyright#7312]
## Environment data

-   Language Server version: PyLance v2024.2.2
-   OS and version: macOS Sonoma 14.3.1
-   Python version: 3.11.7

## Code Snippet

```python
from typing import Iterator

class Leaf:
    def __init__(self, value: float):
        self.value = value

class Node:
    def __init__(self, left: "Tree", right: "Tree"):
        self.left = left
        self.right = right

Tree = Leaf | Node

def visit(tree: Tree, skip=lambda t: False) -> Iterator[Tree]:
    if not skip(tree):
        yield tree
    if not isinstance(tree, Leaf):
        visit(tree.left)
        visit(tree.right)

def sum_tree(tree: Tree) -> float:
    return sum(l.value for l in visit(tree, skip=lambda t: isinstance(t, Node)))
```

The last line is rightly rejected by the type checker, but the diagnosis is wrong:

```
Cannot access member "value" for type "Node"
  Member "value" is unknown
```

Type is actually Leaf, I would be surprised if the static analyzer could get that, but it should at least infer Tree.

```python
def left_trees(tree: Tree) -> list[Tree]:
    return [l.left for l in visit(tree, skip=lambda t: isinstance(t, Leaf))]
```

This fails in the opposite direction:

```
Cannot access member "left" for type "Leaf"
  Member "left" is unknown
```

**Comment by erictraut** [comment]
Pyright (the type checker upon which pylance is built) is working correctly here.

Contrary to the bug title, there is no type inference involved in this sample. Pyright is applying well-defined type evaluation rules (as dictated by the Python typing spec) each step of the way. No inference is needed here.

Here's how the type evaluation rules are applied. You have indicated that the return type of `visit` is `Iterable[Tree]` where `Tree` is a union of `Leaf` and `Node`. That means `l` within the comprehension evaluates to `Leaf | Node`. The expression `l.left` can be evaluated for `Leaf` but fails in the case of `Node`, which is what the diagnostic is telling you.

**Comment by memeplex** [comment]
Ok, I get your explanation, it's just that the wording of the error message is confusing, it seems to imply that the wrong type was inferred. It's not a logical implication, although I assume most people would read `Cannot access member "left" for type "Leaf"` as if the expression were inferred to be of type `Leaf`. But I'm not on most people head, barely in mine.
